zcash_client_sqlite/src/lib.rs

```diff
@@ -558,6 +558,59 @@ impl<'a, P: consensus::Parameters> WalletWrite for DataConnStmtCache<'a, P> {
         })
     }
 
+    #[cfg(feature = "unstable")]
+    fn remove_tx(&mut self, txid: &TxId) -> Result<(), Self::Error> {
+        self.transactionally(|up| {
+            // Find the transaction to be deleted.
+            let tx_ref = up.stmt_select_tx_ref(txid)?;
+
+            // Delete witnesses for received notes that will be deleted.
+            up.wallet_db.conn.execute_named(
+                "DELETE FROM sapling_witnesses WHERE note IN (
+                    SELECT rn.id_note
+                    FROM received_notes rn
+                    WHERE rn.tx = :tx
+                )",
+                &[(":tx", &tx_ref)],
+            )?;
+
+            // Delete received notes that were outputs of the given transaction.
+            up.wallet_db.conn.execute_named(
+                "DELETE FROM received_notes WHERE tx = :tx",
+                &[(":tx", &tx_ref)],
+            )?;
+
+            // Unmine received notes that were inputs to the given transaction.
+            up.wallet_db.conn.execute_named(
+                "UPDATE received_notes SET spent = null WHERE spent = :tx",
+                &[(":tx", &tx_ref)],
+            )?;
+
+            // We don't normally want to delete sent notes, as this can contain data that
+            // is not recoverable from the chain. However, given that the caller wants to
+            // delete the transaction in which these notes were sent, we cannot keep these
+            // database entries around.
+            up.wallet_db
+                .conn
+                .execute_named("DELETE FROM sent_notes WHERE tx = :tx", &[(":tx", &tx_ref)])?;
+
+            // Delete the UTXOs because these are effectively cached and we don't have a
+            // good way of knowing whether they're spent.
+            up.wallet_db.conn.execute_named(
+                "DELETE FROM utxos WHERE spent_in_tx = :tx",
+                &[(":tx", &tx_ref)],
+            )?;
+
+            // Now that it isn't depended on, delete the transaction.
```

Comment:
It occurs to me that this is not necessarily true: it may be that the caller is deleting a transaction with notes that have already been spent in some downstream transaction. I checked locally and the test works fine if I delete the first transaction (that is, the balance is detected only as the change), because there is only a chain of two transactions (so no spent note becomes falsely unspent). I expect that the actual bug could be reproduced (though it would take some work as the currentl test harness can't create three-tx chains).

This was a bug in the Android SDK logic, but not one that I think is reachable, because the deletion logic is only triggered for cancelled transactions (that have only just been created) and expired transactions, neither of which by definition can have outputs that get spent. But we should either fix this before publishing a `zcash_client_sqlite` release, or remove the API.

Reply:
Opened #622.

Reply:
I've now resolved this issue; #622 is now about whether we want to keep the API itself.